packages/alfa-html/src/alphabet.ts

```diff
@@ -237,6 +264,522 @@ const tagName: Pattern = (stream, emit, state) => {
+      return scriptDataEndTagOpen;
+    case Char.ExclamationMark:
+      stream.advance(1);
+      emit({ type: TokenType.Character, data: "<" });
+      emit({ type: TokenType.Character, data: "!" });
+      return scriptDataEscapeStart;
+    default:
+      emit({ type: TokenType.Character, data: "<" });
+      return scriptData;
+  }
+};
+
+/**
+ * 8.2.4.16
+ * @see https://www.w3.org/TR/html/syntax.html#tokenizer-script-data-end-tag-open-state
+ */
+const scriptDataEndTagOpen: Pattern = (stream, emit, state) => {
+  const char = stream.peek(0);
+  if (char === null || !isAlpha(char)) {
+    emit({ type: TokenType.Character, data: "<" });
+    emit({ type: TokenType.Character, data: "/" });
+    return scriptData;
+  }
+
+  state.tag = {
+    type: TokenType.EndTag,
+    name: ""
+  };
+  return scriptDataEndTagName;
+};
+
+/**
+ * 8.2.4.17
+ * @see https://www.w3.org/TR/html/syntax.html#tokenizer-script-data-end-tag-name-state
+ */
+const scriptDataEndTagName: Pattern = (stream, emit, state) => {
+  function isAppropriateEndTagToken(): boolean {
+    return state.tag !== null && state.tag.name === "script";
+  }
+
+  function anythingElse() {
+    emit({ type: TokenType.Character, data: "<" });
+    emit({ type: TokenType.Character, data: "/" });
+    for (let i = 0; i < state.temporaryBuffer.length; i++) {
+      emit({ type: TokenType.Character, data: state.temporaryBuffer[i] });
+    }
+    return scriptData;
+  }
+
+  const char = stream.peek(0);
+  if (char === null) {
+    return anythingElse;
+  }
+
+  switch (char) {
+    case Char.CharacterTabulation:
+    case Char.LineFeed:
+    case Char.FormFeed:
+    case Char.Space:
+      stream.advance(1);
+      if (isAppropriateEndTagToken()) {
+        return beforeAttributeName;
+      }
+      return anythingElse;
+    case Char.Solidus:
+      stream.advance(1);
+      if (isAppropriateEndTagToken()) {
+        return selfClosingStartTag;
+      }
+      return anythingElse;
+    case Char.GreaterThanSign:
+      stream.advance(1);
+      if (isAppropriateEndTagToken()) {
+        emit(state.tag!);
+        return data;
+      }
+      return anythingElse;
+    default:
+      if (isUpperCase(char)) {
+        stream.advance(1);
+        if (state.tag !== null) {
+          state.tag.name += fromCharCode(char).toLowerCase();
+        } else {
+          state.tag = {
+            type: TokenType.EndTag,
+            name: fromCharCode(char).toLowerCase()
+          };
+        }
+        state.temporaryBuffer += fromCharCode(char);
+        break;
+      }
+      if (isLowerCase(char)) {
+        stream.advance(1);
+        if (state.tag !== null) {
+          state.tag.name += fromCharCode(char);
+        } else {
+          state.tag = {
+            type: TokenType.EndTag,
+            name: fromCharCode(char)
+          };
+        }
+        state.temporaryBuffer += fromCharCode(char);
+        break;
+      }
+      return anythingElse;
+  }
+};
```

Comment:
I'd restructure the above to this:

```ts
const scriptDataEndTagName: Pattern = (stream, emit, state) => {
  const char = stream.peek(0);

  switch (char) {
    case Char.CharacterTabulation:
    case Char.LineFeed:
    case Char.FormFeed:
    case Char.Space:
      stream.advance(1);
      if (isAppropriateEndTagToken(state.start!, state.end!)) {
        return beforeAttributeName;
      }
      break;
    ...
    default:
      if (char !== null && isAlpha(char)) {
        ...
        return;
      }
  }

  emit({ type: TokenType.Character, data: "<" });
  emit({ type: TokenType.Character, data: "/" });

  for (let i = 0; i < state.temporaryBuffer.length; i++) {
    emit({ type: TokenType.Character, data: state.temporaryBuffer[i] });
  }

  return scriptData;
};
```

This way, the flow through the function is kept linear.